Handle users leaving the server while we're applying mutes

Kowlin opened the pull request:
### Type

- [x] Bugfix
- [ ] Enhancement
- [ ] New feature

### Description of the changes

Thanks @Sharkytheking. This PR fixes an issue when a member leaves the guild when a mute is being applied.

```
Traceback (most recent call last):
  File "/red-v32/lib/python3.8/site-packages/discord/ext/commands/core.py", line 79, in wrapped
    ret = await coro(*args, **kwargs)
  File "/red-v32/lib/python3.8/site-packages/redbot/cogs/mod/mutes.py", line 257, in guild_mute
    success, issue = await self.mute_user(guild, channel, author, user, audit_reason)
  File "/red-v32/lib/python3.8/site-packages/redbot/cogs/mod/mutes.py", line 422, in mute_user
    await channel.set_permissions(user, overwrite=overwrites, reason=reason)
  File "/red-v32/lib/python3.8/site-packages/discord/abc.py", line 602, in set_permissions
    await http.edit_channel_permissions(self.id, target.id, allow.value, deny.value, perm_type, reason=reason)
  File "/red-v32/lib/python3.8/site-packages/discord/http.py", line 220, in request
    raise NotFound(r, data)
discord.errors.NotFound: 404 NOT FOUND (error code: 10009): Unknown Overwrite

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/red-v32/lib/python3.8/site-packages/discord/ext/commands/bot.py", line 863, in invoke
    await ctx.command.invoke(ctx)
  File "/red-v32/lib/python3.8/site-packages/redbot/core/commands/commands.py", line 631, in invoke
    await super().invoke(ctx)
  File "/red-v32/lib/python3.8/site-packages/discord/ext/commands/core.py", line 1158, in invoke
    await ctx.invoked_subcommand.invoke(ctx)
  File "/red-v32/lib/python3.8/site-packages/discord/ext/commands/core.py", line 728, in invoke
    await injected(*ctx.args, **ctx.kwargs)
  File "/red-v32/lib/python3.8/site-packages/discord/ext/commands/core.py", line 88, in wrapped
    raise CommandInvokeError(exc) from exc
discord.ext.commands.errors.CommandInvokeError: Command raised an exception: NotFound: 404 NOT FOUND (error code: 10009): Unknown Overwrite
```

Kowlin committed:
Handle when users leave a server while we're applying mutes

Goddamn guilds with >100 text channels >.<

Kowlin committed:
Damn you tox.

Kowlin committed:
Damn you Bread.

Bakersbakebread commented on redbot/cogs/mod/mutes.py:
```diff
@@ -422,6 +423,8 @@ class MuteMixin(MixinMeta):
         except discord.Forbidden:
             return False, _(mute_unmute_issues["permissions_issue"])
+        except discord.NotFound:
+            return False, _(mute_unmute_issues["left_guild"])
```
Could this also be if the channel was deleted as well? 🤣

Kowlin replied:
Added error handling for that issue.

Kowlin committed:
Fixed a bug whereby Guild would always return a success

Also handled error checking on NotFound so that the propper handling can be done.

Kowlin committed:
Fixed flake8 compile error (whoops)

TrustyJAID commented on redbot/cogs/mod/mutes.py:
```diff
@@ -251,27 +253,36 @@ class MuteMixin(MixinMeta):
+
+            if success is False:  # Check the latest entry
+                guild_success = False
+                break
```
This will break if the bot is explicitly denied permissions to handle overwrites inside any channels when trying to mute a user rendering the command useless in a majority of properly setup permission servers. It would be better to handle this after attempting all channels and checking teh `mute_success` list for displaying where things went wrong. Or you may change the return from `mute_user` to be more than just a boolean return statement and then handle things based on severity of the error that way.

Jackenmen replied:
Just changing the return value of `mute_user` to `None` when bot gets Forbidden error should suffice, right?

Kowlin replied:
Yes, it would.

TrustyJAID replied:
There is 1 place where the return value is None but not where we want to explicitly stop the funciton from continuing.

TrustyJAID replied:
But the other things to consider is that a server has bot in an admin channel at the top of the server channel list and the bot is explictly denied permission to manage_roles in that channel. In the current way this is setup the entire mute funciton will fail silently without providing the overwrites to subsequent channels which is expected.

Kowlin replied:
You're right in that regard and I suppose I'll redo some stuff later.

Kowlin committed:
Revert "Fixed flake8 compile error (whoops)"

This reverts commit ec8b67234733c6812defaa7ca2abeab3794292ad.

Kowlin committed:
Revert "Fixed a bug whereby Guild would always return a success"

This reverts commit 8d08ac31af4387e6a1757b4a2e97add55f8633c1.

Kowlin committed:
Apply the lost commits that we actually want

Since cherry picking is hard.

Kowlin committed:
Isn't the English language FUN!?

TrustyJAID reviewed (dismissed):
Looks good to me.

Kowlin committed:
*sigh*
